### ISSUE TITLE
Spack Broken after Upgrade to MacOs Big Sur

### ISSUE THREAD

**Issue by jlconlin** [LLNL/spack#21492]
### Steps to reproduce the issue
I recently upgrade MacOS to Big Sur. (Spack was working previously.) Spack was broken in BigSur so I deleted everything Spack related and started over. Spack is still broken.

```console
$ cd $HOME
$ rm -rf spack/ .spack/
$ git clone https://github.com/spack/spack.git
$ source spack/share/spack/setup-env.sh
$ spack install libelf                                                                                                             
==> Installing libelf-0.8.13-4ifw3i42jvwbag67hn5ylugghq2kmrnh
==> No binary for libelf-0.8.13-4ifw3i42jvwbag67hn5ylugghq2kmrnh found: installing from source
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/59/591a9b4ec81c1f2042a97aa60564e0cb79d041c52faa7416acb38bc95bd2c76d.tar.gz
#################################################################################################################################################### 100.0%
==> libelf: Executing phase: 'autoreconf'
==> libelf: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-libelf-0.8.13-4ifw3i42jvwbag67hn5ylugghq2kmrnh/spack-src/configure' '--prefix=/Users/jlconlin/spack/opt/spack/darwin-bigsur-skylake/apple-clang-12.0.0/libelf-0.8.13-4ifw3i42jvwbag67hn5ylugghq2kmrnh' '--enable-shared' '--disable-dependency-tracking' '--disable-debug'

1 error found in build log:
     33    checking size of int... 0
     34    checking size of long... 0
     35    checking size of long long... 0
     36    checking size of __int64... 0
     37    checking for 64-bit integer... no
     38    checking for 32-bit integer... no
  >> 39    configure: error: neither int nor long is 32-bit

See build log for details:
  /var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-libelf-0.8.13-4ifw3i42jvwbag67hn5ylugghq2kmrnh/spack-build-out.txt
```

`libelf` was not the only package that was struggling to install. I also tried installing `cmake`; it failed, but for different reasons.

```console
$ spack install cmake                                                                                                               
==> Installing pkgconf-1.7.3-h7h25xohkjm36qltjovbxhgrpb7s5ett
==> No binary for pkgconf-1.7.3-h7h25xohkjm36qltjovbxhgrpb7s5ett found: installing from source
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/b8/b846aea51cf696c3392a0ae58bef93e2e72f8e7073ca6ad1ed8b01c85871f9c0.tar.xz
#################################################################################################################################################### 100.0%
==> pkgconf: Executing phase: 'autoreconf'
==> pkgconf: Executing phase: 'configure'
==> pkgconf: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'
4 errors found in build log:
     114    config.status: creating tests/test_env.sh
     115    config.status: creating libpkgconf/config.h
     116    config.status: executing depfiles commands
     117    config.status: executing libtool commands
     118    ==> pkgconf: Executing phase: 'build'
     119    ==> [2021-02-04-11:02:40.703748] 'make' '-j16'
  >> 120    make: /var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-pkgconf-1.7.3-h7h25xohkjm36qltjovbxhgrpb7s5ett/spack-s
            rc/install-sh: Permission denied
  >> 121    make: /var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-pkgconf-1.7.3-h7h25xohkjm36qltjovbxhgrpb7s5ett/spack-s
            rc/install-sh: Permission denied
  >> 122    make: *** [libpkgconf/.deps/.dirstamp] Error 1
     123    make: *** Waiting for unfinished jobs....
  >> 124    make: *** [libpkgconf/.dirstamp] Error 1
See build log for details:
  /var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-pkgconf-1.7.3-h7h25xohkjm36qltjovbxhgrpb7s5ett/spack-build-out.txt
==> Warning: Skipping build of ncurses-6.2-gikh646beft7ztei3wiekwymc4dtwv43 since pkgconf-1.7.3-h7h25xohkjm36qltjovbxhgrpb7s5ett failed
==> Warning: Skipping build of readline-8.0-cfoqvf3uxnuhf6wjwvqz3c3qgshzuhgq since ncurses-6.2-gikh646beft7ztei3wiekwymc4dtwv43 failed
==> Warning: Skipping build of gdbm-1.18.1-mhckngfgvmlilgxlnchozhfsem5putka since readline-8.0-cfoqvf3uxnuhf6wjwvqz3c3qgshzuhgq failed
==> Warning: Skipping build of perl-5.32.1-fyr7pumixy24t4iickul3i7iugar2arv since gdbm-1.18.1-mhckngfgvmlilgxlnchozhfsem5putka failed
==> Warning: Skipping build of openssl-1.1.1i-csloafmhlbbqanwqgc6jmwo6gx7ffvjm since perl-5.32.1-fyr7pumixy24t4iickul3i7iugar2arv failed
==> Warning: Skipping build of cmake-3.19.2-ur6ousaun7fyz3b2oaf3l34iiorm2p2r since openssl-1.1.1i-csloafmhlbbqanwqgc6jmwo6gx7ffvjm failed
==> Installing berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm
==> No binary for berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm found: installing from source
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/0c/0cecb2ef0c67b166de93732769abdeba0555086d51de1090df325e18ee8da9c8.tar.gz
#################################################################################################################################################### 100.0%
==> berkeley-db: Executing phase: 'autoreconf'
==> berkeley-db: Executing phase: 'configure'
==> berkeley-db: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'
4 errors found in build log:
     381    libtool: compile:  /Users/jlconlin/spack/lib/spack/env/clang/clang -c -I. -I/private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/
            spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src -O3 /private/var/folders/5t/txmxvbk97018dv7zj6mxqd1
            80002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src/btree/bt_rsearch.c  -fno-common
            -DPIC -o .libs/bt_rsearch.o
     382    libtool: compile:  /Users/jlconlin/spack/lib/spack/env/clang/clang -c -I. -I/private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/
            spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src -O3 /private/var/folders/5t/txmxvbk97018dv7zj6mxqd1
            80002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/lang/db185/db185.c  -fno-common -DPI
            C -o .libs/db185.o
     383    libtool: compile:  /Users/jlconlin/spack/lib/spack/env/clang/clang -c -I. -I/private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/
            spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src -O3 /private/var/folders/5t/txmxvbk97018dv7zj6mxqd1
            80002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src/btree/bt_recno.c  -fno-common -D
            PIC -o .libs/bt_recno.o
     384    libtool: compile:  /Users/jlconlin/spack/lib/spack/env/clang/clang -c -I. -I/private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/
            spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src -O3 /private/var/folders/5t/txmxvbk97018dv7zj6mxqd1
            80002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src/btree/bt_put.c  -fno-common -DPI
            C -o .libs/bt_put.o
     385    libtool: compile:  /Users/jlconlin/spack/lib/spack/env/clang/clang -c -I. -I/private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/
            spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src -O3 /private/var/folders/5t/txmxvbk97018dv7zj6mxqd1
            80002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src/btree/bt_compress.c  -fno-common
             -DPIC -o .libs/bt_compress.o
     386    ./libtool --mode=compile /Users/jlconlin/spack/lib/spack/env/clang/clang -c -I. -I/private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlc
            onlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src  -O3  /private/var/folders/5t/txmxvbk97018dv7
            zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src/btree/bt_split.c
  >> 387    make: ./libtool: Permission denied
     388    ./libtool --mode=compile /Users/jlconlin/spack/lib/spack/env/clang/clang -c -I. -I/private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlc
            onlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src  -O3  /private/var/folders/5t/txmxvbk97018dv7
            zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-src/src/btree/bt_stat.c
  >> 389    make: ./libtool: Permission denied
  >> 390    make: *** [bt_split.lo] Error 1
     391    make: *** Waiting for unfinished jobs....
  >> 392    make: *** [bt_stat.lo] Error 1
     393    /private/var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm
            /spack-src/lang/db185/db185.c:71:7: warning: case value not in enumerated type 'DBTYPE' [-Wswitch]
     394            case 0:                                 /* DB_BTREE */
     395                 ^
     396    1 warning generated.
See build log for details:
  /var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-berkeley-db-18.1.40-ylfrfleejl6iutbxwg2kokkajcq6tqrm/spack-build-out.txt
==> Installing zlib-1.2.11-xflsouiaysa3uizi6mjcwhv54aa5m6kd
==> No binary for zlib-1.2.11-xflsouiaysa3uizi6mjcwhv54aa5m6kd found: installing from source
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/c3/c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1.tar.gz
#################################################################################################################################################### 100.0%
==> zlib: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'
4 errors found in build log:
     22    /Users/jlconlin/spack/lib/spack/env/clang/clang -fPIC -O2 -DHAVE_HIDDEN  -c -o inflate.o inflate.c
     23    /Users/jlconlin/spack/lib/spack/env/clang/clang -fPIC -O2 -DHAVE_HIDDEN  -c -o inftrees.o inftrees.c
     24    /Users/jlconlin/spack/lib/spack/env/clang/clang -fPIC -O2 -DHAVE_HIDDEN  -c -o trees.o trees.c
     25    /Users/jlconlin/spack/lib/spack/env/clang/clang -fPIC -O2 -DHAVE_HIDDEN  -c -o zutil.o zutil.c
     26    /Users/jlconlin/spack/lib/spack/env/clang/clang -fPIC -O2 -DHAVE_HIDDEN  -c -o compress.o compress.c
     27    /Users/jlconlin/spack/lib/spack/env/clang/clang -fPIC -O2 -DHAVE_HIDDEN  -c -o uncompr.o uncompr.c
  >> 28    make: /Users/jlconlin/spack/lib/spack/env/clang/clang: Permission denied
     29    /Users/jlconlin/spack/lib/spack/env/clang/clang -fPIC -O2 -DHAVE_HIDDEN  -c -o gzclose.o gzclose.c
  >> 30    make: /Users/jlconlin/spack/lib/spack/env/clang/clang: Permission denied
  >> 31    make: *** [uncompr.o] Error 1
     32    make: *** Waiting for unfinished jobs....
  >> 33    make: *** [gzclose.o] Error 1
See build log for details:
  /var/folders/5t/txmxvbk97018dv7zj6mxqd180002jq/T/jlconlin/spack-stage/spack-stage-zlib-1.2.11-xflsouiaysa3uizi6mjcwhv54aa5m6kd/spack-build-out.txt
```

### Error Message

```console
$ spack --debug --stacktrace <command>
```

### Information on your system

```console
$ spack debug report                                                                                                                
* **Spack:** 0.16.0-1133-ba70f90ee0
* **Python:** 3.8.3
* **Platform:** darwin-bigsur-skylake
* **Concretizer:** original
```

The only configuration file I have is in `$HOME/.spack/darwin/compilers.yaml`, which I did not create myself.
```console
compilers:
- compiler:
    spec: apple-clang@12.0.0
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77:
      fc:
    flags: {}
    operating_system: bigsur
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

### Additional information

- [x ] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [ x] I have searched the issues of this repo and believe this is not a duplicate
- [x ] I have run the failing commands in debug mode and reported the output

**Comment by sethrj** [comment]
I think libelf has always been a problem for macOS? https://github.com/spack/spack/issues/8401

**Comment by sethrj** [comment]
@jlconlin It looks like the `clang` copy or alias is broken: all the failures say
```
/Users/jlconlin/spack/lib/spack/env/clang/clang: Permission denied
```
Could you see what that file is, whether it's a symlink to a nonexistent exe, etc? I also assume you've got xcode and the command line tools installed?

**Comment by jlconlin** [comment]
> I think libelf has always been a problem for macOS? #8401

It doesn't really matter what I try to install, it always barfs on me.

**Comment by jlconlin** [comment]
> @jlconlin It looks like the `clang` copy or alias is broken: all the failures say
> 
> ```
> /Users/jlconlin/spack/lib/spack/env/clang/clang: Permission denied
> ```
> 
> Could you see what that file is, whether it's a symlink to a nonexistent exe, etc? I also assume you've got xcode and the command line tools installed?

The clang alias you suggest points o the executable `cc` one directory up. Here is what that directory looks like for me:
```
09:13:19 mickey ~/spack/lib/spack/env$ ll
total 32
drwxr-xr-x  5 jlconlin  staff   160B Mar 16  2020 arm
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 c++ -> cc
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 c89 -> cc
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 c99 -> cc
drwxr-xr-x  3 jlconlin  staff    96B Mar 16  2020 case-insensitive
-rwxr-xr-x  1 jlconlin  staff    16K Mar 16  2020 cc
drwxr-xr-x  5 jlconlin  staff   160B Mar 16  2020 cce
drwxr-xr-x  6 jlconlin  staff   192B Mar 16  2020 clang
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 cpp -> cc
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 f77 -> cc
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 f90 -> cc
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 f95 -> cc
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 fc -> cc
drwxr-xr-x  5 jlconlin  staff   160B Mar 16  2020 fj
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 ftn -> cc
drwxr-xr-x  5 jlconlin  staff   160B Mar 16  2020 gcc
drwxr-xr-x  5 jlconlin  staff   160B Mar 16  2020 intel
lrwxr-xr-x  1 jlconlin  staff     2B Mar 16  2020 ld -> cc
drwxr-xr-x  3 jlconlin  staff    96B Mar 16  2020 nag
drwxr-xr-x  5 jlconlin  staff   160B Mar 16  2020 pgi
drwxr-xr-x  6 jlconlin  staff   192B Mar 16  2020 xl
drwxr-xr-x  6 jlconlin  staff   192B Mar 16  2020 xl_r
09:13:21 mickey ~/spack/lib/spack/env$ ./cc
Spack compiler must be run from Spack! Input 'SPACK_ENV_PATH' is missing.
```
It's confusing to me.

**Comment by cjosey** [comment]
I have a computer that should be configured identically to @jlconlin 's, and I upgraded to 11.4 today and had the same issue. I was able to partially track it down and confirm it isn't a Spack-specific issue.  Building `pkgconf` 100% outside of Spack results in a file permissions error if it is the first run and `-jx` (`x`>1) is used.

After some testing, the solutions I found were:
1. Disable parallel building (`spack install -j1 cmake` or set `build_jobs` in the configure files).
2. Build twice (modifying Spack to try once and try again).  Not practical.
3. Install `gmake` and load it (`spack install -j1 gmake` followed by `spack load gmake`)

I do not know if this is a general issue with macOS 11's GNU Make, or if it's a problem with our specific configuration.  As I do not have a macOS machine configured normally, I cannot disambiguate.

If others find this issue, it may be worth documenting, but I think it may be specific to us.
